### ISSUE TITLE
Non-destructive platform update

### ISSUE THREAD

**Issue by purplecabbage** [cordova/cordova-discuss#50]
Although we designed/intended for platforms to be build artifacts, there will always be cases where the app developer has made changes that they would like to keep.  
Currently if an app developer runs `cordova platform update ios` any changes they have made to the xcode project, or any files present in the new template will be overwritten.

While being aware of and migrating any changes is absolutely out of scope, it would great if we could keep the older version around so that developers could at least have the chance to do their own migration.
# Proposal:

_(using ios for example only, this applies to all platforms)_

Assuming the app developer has already done a `cordova platform add ios@4.0.0` 
Their file structure then looks like this:

```
./platforms/ios/
```

If they choose to update, running `cordova platform update ios`
We will:
- rename the current ios to ios@4.0.0 based on it's own version number
- fetch the latest ios as we do already, and place it in `./platforms/ios`

So the developer's file structure then looks like this:

```
./platforms/ios/
./platforms/ios@4.0.0
```

This has been discussed in the past, and I believe there is consensus that this is a good idea.  I just wanted to formalize it here before we move forward with it.

Some out of scope nice-to-have's that will be a part of another proposal:
- allow running of each platform@version, so the dev could still run
  - `cordova platform run ios@4.0.0`
- support downgrading (which could be destructive?)
  - `cordova platform downgrade|revert ios`
- support adding older platform versions
  - `cordova platform add ios@4.0.0 # When platforms/ios/ already has v4.1.1`

**Comment by kerrishotts** [comment]
Will the update process inform the dev that it's created the ios@4.0.0 directory and that they may need to do manual migration? 

Also, if I'm absolutely certain about "ios" being a build artifact (which is the case for most of my projects), is there a way to opt out of this behavior (or is the intent for this case simply to "rm" and "add")?

**Comment by brodybits** [comment]
I am also concerned this could lead to very hard to reproduce bug reports, especially with some of the more complex plugins.

**Comment by purplecabbage** [comment]
@kerrishotts Yes, the process should inform the developer that their previous project is still available and Yes, there could also be a flag to get the current behavior.  --replace or something.

@brodybits The potential for plugin issues is no different than it is currently with rm/add. It is not until we get to the out of scope items that things get mucky, which is why I consider them out of scope for now; they need further exploration.  I don't believe the choices I have outlined will in any way force how we proceed.

**Comment by imhotep** [comment]
Is there going to be a limit to the number of projects laying around ? I would think that the right way to do this is to have every folder named as `platform@version`and have a `platform` symbolic link pointing to the right one (kinda like JRE/JSE). Do Symbolic Links work on Windows though ?

**Comment by macdonst** [comment]
I'm not in favour of this idea. I continue to like the idea that the platforms directory is just a build artifact. However, people needing to go into the platforms directory and make some changes is a reality. Do we know the use cases as to why people need to do this? I would prefer to surface those use cases and enhance Cordova so that you don't have to mess around with the platforms directory.

**Comment by purplecabbage** [comment]
@imhotep Yes, I thought of having a platform/latest that was a symlink but that would require changes to every platform command. 
symlinks should be avoided at all costs, as they require admin privileges on Windows (~50% of cli users )

**Comment by purplecabbage** [comment]
@macdonst I love your optimism, however I feel it is unobtainable. This should continue to be our lofty goal that we chase, but we cannot expect to prettily abstract every nuance of app submission and project configuration for every platform, present and future. 
This solution is saying "Here is a category of issues we are currently ignoring that we can at least provide a workable solution for. (and quite easily)"
You seem to be saying "Let's just fix all the problems including the ones we don't yet know exist."

Of course we can continue towards fixing everything for everyone, but I see no reason why that endeavor needs to block this proposal.

**Comment by macdonst** [comment]
@purplecabbage I'm not sure if you meant for your comment to me to have a condescending tone but that is the way it came across to me. I'm going to assume that is not in the spirit in which you meant it though.

As I said in my earlier comment, I understand that there is a problem here. I'm just not sure that this is the best solution without understanding in which situations people need to modify the platforms folder. This is why I asked what are the use cases we are trying to solve here.

@ktop's recent changes to the CLI to allow plugin.xml and config.xml to edit configuration files is a great step forward to not needing to modify the platforms folder. If we could ever get https://github.com/cordova/cordova-discuss/issues/6 implemented as a way to deliver not just images but arbitrary resource files to the platforms directory that would solve another set of problems. If changes to the Xcode project are getting over-written then maybe we need to add the config-file tag into config.xml in order to allow the app to change the plist without any magic on our side.

Also, a non-destructive platform update doesn't help downstream build services like PhoneGap Build and IntelXDK as those users are be locked out of being able to modify anything in the platforms directory.

Regardless, I never said that the proposal should be blocked only that I'm in favour of continuing to treat the platform folders as artifacts.

**Comment by imhotep** [comment]
@macdonst one problem that I faced not long ago is enabling swift support for an iOS project (if you want to use swift plugins for example). Not sure how much work is required to have this supported in `config.xml`/`plugin.xml`. There are probably other XCode related settings that are not supported and probably will never be supported.

**Comment by purplecabbage** [comment]
I meant no disrespect, and I am sorry it came across that way.

A few iOS things that are hard to change without modifying the plist directly:
Localized display names, bundle identifier, updating bundle version to update an existing app in the app store, url types, ...
There is likely a similar list for android and windows and abstracting them all neatly would take a huge effort.

IMHO, we should not concern ourselves with downstream build services.  We can do our best to accommodate them, but we cannot let them inhibit our progress.

We have done and continue to do work to help developers who wish to mix some native development with hybrid development.  Consider a case where instead of adding cordova to an existing app, a developer starts with a cordova app, and adds more native views to it.  If they ever want to upgrade the cordova bits they currently would have to create a new cordova app, and re-apply their changes.  

We don't have to stop treating platforms as build artifacts, but we should stop forcing developers to do so if it is reasonable to simply not delete some stuff. That, and that alone is my entire goal with this proposal.

re: ... I never said that the proposal should be blocked ...
Your first statement was 'I'm not in favour of this idea.' which I took as blocking.

**Comment by goya** [comment]
just add a --backup flag to it if people want it. 

I hope I sound condescending, overly-optimistic, dreamy, uninformed and thin-skinned all at once! you guys are hilarious!

**Comment by purplecabbage** [comment]
Prior art, we actually discussed this and got to consensus previously.  
http://markmail.org/message/nrabt2f4kcbp6vkh

**Comment by shazron** [comment]
So is the versioning going to be like nvm? That we can switch versions?

**Comment by purplecabbage** [comment]
@shazron Maybe one day, but I have marked that as out of scope.
The intent here is just to not destroy any changes they may have made, and allow app developers to migrate their own changes, or delete the archive themselves.

https://github.com/apache/cordova-lib/pull/465 implements the full scope of this discuss as far as I intended.

**Comment by purplecabbage** [comment]
closing, ready for 7
